cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,244 @@
+#include "arrow/util/macros.h"
+#include "arrow/util/string_view.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+class MemoryPool;
+
+using internal::checked_cast;
+using internal::DictionaryTraits;
+using internal::HashTraits;
+
+namespace compute {
+
+#define CHECK_IMPLEMENTED(KERNEL, FUNCNAME, TYPE)                                       \
+  if (!KERNEL) {                                                                        \
+    return Status::NotImplemented(FUNCNAME, " not implemented for ", type->ToString()); \
+  }
+
+class HashBinaryKernelImpl: public BinaryKernel {
+
+    virtual Status Compute(FunctionContext* ctx, const ArrayData& left,
+                         const ArrayData& right, std::shared_ptr<ArrayData>& out) = 0;
+
+    public:
+    Status Call(FunctionContext* ctx, const Datum& left, const Datum& right,
+                                                    Datum* out) override {
+
+        DCHECK_EQ(Datum::ARRAY, right.kind());
+        DCHECK_EQ(Datum::ARRAY, left.kind());
+
+        const ArrayData& left_data = *left.array();
+        const ArrayData& right_data = *right.array();
+
+        std::shared_ptr<ArrayData> result;
+        RETURN_NOT_OK(Compute(ctx, left_data , right_data, result));
+
+        out->value = std::move(result);
+        return Status::OK();
+    }
+    std::shared_ptr<DataType> out_type() const override { return boolean(); }
+};
+
+template <typename Type, typename Scalar>
+class IsInKernel : public HashBinaryKernelImpl{
+public:
+    IsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool){}
+
+    Status VisitNull(){
+        values.emplace_back ();
+        return Status::OK();
+    }
+
+    Status VisitValue(const Scalar& value){
+        values.emplace_back(value);
+        return Status::OK();
+    }
+
+    Status Compute(FunctionContext* ctx, const ArrayData& left, const ArrayData& right,
+                 std::shared_ptr<ArrayData>& out) override {
+
+        values.reserve(left.length);
+        memo_table_.reset(new MemoTable(0));
+        bool_builder_.Reset();
+        bool_builder_.Reserve(left.length);
+
+        ArrayDataVisitor<Type>::Visit(right, this);
+
+        for (const auto& value: values){
+            memo_table_->GetOrInsert(value);
+        }
+
+        values.clear();
+        ArrayDataVisitor<Type>::Visit(left, this);
+
+        for (const auto& value: values){
+            if ( memo_table_->Get(value) != -1){
+                bool_builder_.UnsafeAppend(true);
+            } else {
+                bool_builder_.UnsafeAppend(false);
+            }
+        }
+
+        RETURN_NOT_OK (bool_builder_.FinishInternal(&out));
+        return Status::OK();
+    }
+
+    protected:
+        using MemoTable = typename HashTraits<Type>::MemoTableType;
+        std::unique_ptr<MemoTable> memo_table_;
+
+    private:
+        BooleanBuilder bool_builder_;
+        std::vector<Scalar>values;
+};
+
+
+class NullIsInKernel : public HashBinaryKernelImpl{
```

Comment:
I think it should be possible to collapse this into one kernel.

I think the trick is you probably need two visitors.  One that populate the memo table + an additional member to keep track of nulls on the kernel above.  The second visitor would make use of the memo table and iterate over left.